community/kernel-api/src/test/java/org/neo4j/internal/kernel/api/RelationshipTransactionStateTestBase.java

```diff
@@ -358,6 +361,247 @@ public void shouldSeeNewRelationshipPropertyInTransaction() throws Exception
+            {
+                tx.dataRead().singleRelationship( relationshipId, relationship );
+                assertTrue( "should access relationship", relationship.next() );
+
+                relationship.properties( property );
+                assertTrue( property.next() );
+                assertEquals( propToken, property.propertyKey() );
+                assertEquals( property.propertyValue(), stringValue( "hello" ) );
+
+                assertFalse( "should only find one properties", property.next() );
+                assertFalse( "should only find one relationship", relationship.next() );
+            }
+
+            tx.success();
+        }
+
+        try ( org.neo4j.graphdb.Transaction ignored = graphDb.beginTx() )
+        {
+            assertThat(
+                    graphDb.getRelationshipById( relationshipId ).getProperty( propKey ), equalTo( "hello" ) );
+        }
+    }
+
+    @Test
+    public void shouldSeeAddedPropertyFromExistingRelationshipWithPropertiesInTransaction() throws Exception
+    {
+        // Given
+        long relationshipId;
+        String propKey1 = "prop1";
+        String propKey2 = "prop2";
+        int propToken1;
+        int propToken2;
+        try ( Transaction tx = session.beginTransaction() )
+        {
+            Write write = tx.dataWrite();
+            relationshipId = write.relationshipCreate( write.nodeCreate(),
+                    tx.tokenWrite().relationshipTypeGetOrCreateForName( "R" ), write.nodeCreate() );
+            propToken1 = session.token().propertyKeyGetOrCreateForName( propKey1 );
+            assertEquals( write.relationshipSetProperty( relationshipId, propToken1, stringValue( "hello" ) ), NO_VALUE );
+            tx.success();
+        }
+
+        // When/Then
+        try ( Transaction tx = session.beginTransaction() )
+        {
+            propToken2 = session.token().propertyKeyGetOrCreateForName( propKey2 );
+            assertEquals( tx.dataWrite().relationshipSetProperty( relationshipId, propToken2, stringValue( "world" ) ), NO_VALUE );
+
+            try ( RelationshipScanCursor relationship = cursors.allocateRelationshipScanCursor();
+                  PropertyCursor property = cursors.allocatePropertyCursor() )
+            {
+                tx.dataRead().singleRelationship( relationshipId, relationship );
+                assertTrue( "should access relationship", relationship.next() );
+
+                relationship.properties( property );
+
+                //property 2, start with tx state
+                assertTrue( property.next() );
+                assertEquals( propToken2, property.propertyKey() );
+                assertEquals( property.propertyValue(), stringValue( "world" ) );
+
+                //property 1, from disk
+                assertTrue( property.next() );
+                assertEquals( propToken1, property.propertyKey() );
+                assertEquals( property.propertyValue(), stringValue( "hello" ) );
```

Comment:
Does this imply that the properties contained in the transaction state should always be siphoned out before the persistent state?